### PR TITLE
z_kaleido_collect.c (1 non-matching) and Mostly Documented (Pause Menu Quest Page) 

### PR DESCRIPTION
**Pull request by engineer124**
I avoided completely documenting this file due to the non-matching, but I did name all the data. 

Here's the non-matching, it's very close:
https://decomp.me/scratch/kf3PR

### CODE REVIEW

File: src/code/z_parameter.c [hunk 1]
```diff
@@ -1172,7 +1172,9 @@ u8 Item_Give(PlayState* play, u8 item) {
     }
 
     if (item == ITEM_SKULL_TOKEN) {
-        SET_QUEST_ITEM(item - ITEM_SKULL_TOKEN + QUEST_SKULL_TOKEN);
+        //! @bug: Sets QUEST_QUIVER instead of QUEST_SKULL_TOKEN
+        // Setting `QUEST_SKULL_TOKEN` will result in misplace digits on the pause menu - quest page.
```
Comment (EllipticEllipsis) [inline src/code/z_parameter.c:1176]:
```suggestion
        // Setting `QUEST_SKULL_TOKEN` will result in misplaced digits on the pause menu - Quest page.
```

File: src/overlays/actors/ovl_En_Gakufu/z_en_gakufu.c [hunk 2]
```diff
@@ -115,9 +115,12 @@ void EnGakufu_ProcessNotes(EnGakufu* this) {
 
     songIndex = this->songIndex;
     ocarinaSongButtons = &gOcarinaSongButtons[songIndex];
+
+    //! FAKE:
+    if (1) {}
     songNumButtons = gOcarinaSongButtons[this->songIndex].numButtons;
 
-    for (i = 0; i < songNumButtons; i++) {
+    for (i = 0; i < (u8)songNumButtons; i++) {
```
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_En_Gakufu/z_en_gakufu.c:123]: What happened here?

File: src/overlays/kaleido_scope/ovl_kaleido_scope/z_kaleido_collect.c [hunk 3]
```diff
@@ -1,103 +1,1003 @@
...
+        255, // QUEST_SONG_EPONA
+        255, // QUEST_SONG_SOARING
+        255, // QUEST_SONG_STORMS
+        255, // QUEST_SONG_SUN
+    };
+    static TexturePtr sQuestUpgradeTextures[][3] = {
+        { 0x08053000, 0x08054000, 0x08055000 }, // UPG_QUIVER
+        { 0x08056000, 0x08057000, 0x08058000 }, // UPG_BOMB_BAG
+    };
+    static u8 sQuestUpgrades[] = { UPG_QUIVER, UPG_BOMB_BAG };
```
Comment (EllipticEllipsis) [inline src/overlays/kaleido_scope/ovl_kaleido_scope/z_kaleido_collect.c:126]: I have to ask: all needed in the function?

File: src/overlays/kaleido_scope/ovl_kaleido_scope/z_kaleido_collect.c [hunk 4]
```diff
@@ -1,103 +1,1003 @@
...
+                    pauseCtx->ocarinaStaff->pos = 0;
+                    pauseCtx->ocarinaStaff->state = 0xFE;
+                    pauseCtx->unk_200 = 5;
+                }
+            }
+        }
+    }
+
+    // Draw Skull Token Count
+    // QUEST_SKULL_TOKEN never properly set, see `Item_Give`,
```
Comment (EllipticEllipsis) [inline src/overlays/kaleido_scope/ovl_kaleido_scope/z_kaleido_collect.c:498]:
```suggestion
    // QUEST_SKULL_TOKEN never properly set, see Item_Give(),
```
Think this gets Doxygen to link it automatically.

File: src/overlays/kaleido_scope/ovl_kaleido_scope/z_kaleido_collect.c [hunk 5]
```diff
@@ -1,103 +1,1003 @@
...
+    /* 0x0 */ s16 up;
+    /* 0x2 */ s16 down;
+    /* 0x4 */ s16 left;
+    /* 0x6 */ s16 right;
+} CursorPointDirection;
 
-s16 D_8082B024 = 1;
+void KaleidoScope_UpdateQuestCursor(PlayState* play) {
+    static s16 sQuestSongPlaybackDelayTimer = 0;
+    static CursorPointDirection sCursorPointLinks[] = {
```
Comment (EllipticEllipsis) [inline src/overlays/kaleido_scope/ovl_kaleido_scope/z_kaleido_collect.c:595]: Here too?

File: src/overlays/kaleido_scope/ovl_kaleido_scope/z_kaleido_collect.c [hunk 6]
```diff
@@ -1,103 +1,1003 @@
...
+                        sQuestSongPlaybackDelayTimer = 10;
+
+                        for (i = 0; i < 8; i++) {
+                            sQuestSongPlayedOcarinaButtons[i] = OCARINA_BTN_INVALID;
+                            sQuestSongPlayedOcarinaButtonsAlpha[i] = 0;
+                        }
+
+                        sQuestSongPlayedOcarinaButtonsNum = 0;
+
+                        // Have the player play a song, immediately cancelled below
```
Comment (EllipticEllipsis) [inline src/overlays/kaleido_scope/ovl_kaleido_scope/z_kaleido_collect.c:817]: I suppose this means the audio player? This is the "song demo" mode, I assume. Actually reading below I'm not so sure... can I ask that when it's the audio one it's always called `audio player`, and when it's Player player, it's `Player`?

File: src/overlays/actors/ovl_En_Gakufu/z_en_gakufu.c (hunk 2 above)
Comment (engineer124) [reply src/overlays/actors/ovl_En_Gakufu/z_en_gakufu.c:123]: This was the result of changing:
`u32 AudioOcarina_SetInstrument(u8 ocarinaInstrumentId);`
to
`void AudioOcarina_SetInstrument(u8 ocarinaInstrumentId);`

The latter should be the correct one, and indeed `void` is required to match these quest functions. However, changing that function to `u32` was a trick to match that gakufu function, but it seemed like a fake match to begin with. An alternative match was found by Anon (huge thanks to them), but presumably there's something better out there. Here's the conversation where I mention it:

https://discord.com/channels/688807550715560050/688851997352263820/1013314490625949718

File: src/overlays/kaleido_scope/ovl_kaleido_scope/z_kaleido_collect.c (hunk 3 above)
Comment (engineer124) [reply src/overlays/kaleido_scope/ovl_kaleido_scope/z_kaleido_collect.c:126]: Unfortunately yes. All those color pulsations tend to use data, and those all tend to be directly affected by in-function data. So `sQuestRemainsColorTimer` and those color temps at the top need to be in-function, meaning everything below does as well

File: src/overlays/kaleido_scope/ovl_kaleido_scope/z_kaleido_collect.c (hunk 5 above)
Comment (engineer124) [reply src/overlays/kaleido_scope/ovl_kaleido_scope/z_kaleido_collect.c:595]: Yeah, this one is very unfortunate. `sQuestSongPlaybackDelayTimer` specifically affects regalloc, and `sCursorPointLinks` does to maintain ordering

File: src/overlays/kaleido_scope/ovl_kaleido_scope/z_kaleido_collect.c (hunk 3 above)
Comment (EllipticEllipsis) [reply src/overlays/kaleido_scope/ovl_kaleido_scope/z_kaleido_collect.c:126]: 😮‍💨

File: src/overlays/kaleido_scope/ovl_kaleido_scope/z_kaleido_collect.c (hunk 6 above)
Comment (engineer124) [reply src/overlays/kaleido_scope/ovl_kaleido_scope/z_kaleido_collect.c:817]: I can rephrase this to not use player. It's more "user input" i.e. you giving ocarina inputs, to distinguish it from the game playing back the song to you on auto-pilot.

File: src/overlays/kaleido_scope/ovl_kaleido_scope/z_kaleido_collect.c (hunk 5 above)
Comment (EllipticEllipsis) [reply src/overlays/kaleido_scope/ovl_kaleido_scope/z_kaleido_collect.c:595]: Well I *suppose* you could forward-declare it, but no thanks.

File: src/overlays/kaleido_scope/ovl_kaleido_scope/z_kaleido_collect.c (hunk 6 above)
Comment (EllipticEllipsis) [reply src/overlays/kaleido_scope/ovl_kaleido_scope/z_kaleido_collect.c:817]: Oh right, yeah it's not very clear atm.

Comment (AngheloAlf) [reply src/overlays/kaleido_scope/ovl_kaleido_scope/z_kaleido_collect.c:817]: Agreed, this needs a rewording

File: src/overlays/kaleido_scope/ovl_kaleido_scope/z_kaleido_collect.c [hunk 7]
```diff
@@ -1,103 +1,1003 @@
...
+        }
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_kaleido_scope/func_80817B5C.s")
+s32 KaleidoScope_UpdateQuestStatusPoint(PauseContext* pauseCtx, s16 point) {
+    pauseCtx->cursorPoint[PAUSE_QUEST] = point;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_kaleido_scope/func_80818904.s")
+    return 1;
```
Comment (hensldm) [inline src/overlays/kaleido_scope/ovl_kaleido_scope/z_kaleido_collect.c:1002]: Any reason not to return `true` here?

**Review by hensldm — APPROVED** [review]
1 food for thought

**Review by AngheloAlf — APPROVED** [review]

File: src/overlays/kaleido_scope/ovl_kaleido_scope/z_kaleido_collect.c [hunk 8]
```diff
@@ -1,103 +1,1005 @@
...
+    };
+    static u8 sQuestUpgrades[] = { UPG_QUIVER, UPG_BOMB_BAG };
+    PauseContext* pauseCtx = &play->pauseCtx;
+    s16 sp1CA; // colorSetR and numOcarinaButtons
+    s16 sp1C8; // colorSetG and ocarinaButtonIndex
+    s16 sp1C6;
+    s16 sp1C4;
+    s16 var_v1;
+    s16 i;
+    s16 j;
```
Comment (EllipticEllipsis) [inline src/overlays/kaleido_scope/ovl_kaleido_scope/z_kaleido_collect.c:134]: Is `j` always the vtx index? And would it make sense for `var_s3` to be `k`?

Comment (engineer124) [reply src/overlays/kaleido_scope/ovl_kaleido_scope/z_kaleido_collect.c:134]: No unfortunately, `j` is reused at the end for indexing the digit.

`var_s3` -> `k` is fine though

**Review by EllipticEllipsis — APPROVED** [review]
